test-vectors.md

```diff
@@ -71,81 +71,115 @@ Verification:
-          "ciphertext": /* hex-encoded binary data */
-        },
-        ...
-      ]
-    }
+    [
+      {
+        "generation": /* uint32 */
+        "handshake_key": /* hex-encoded binary data */,
+        "handshake_nonce": /* hex-encoded binary data */,
+        "application_key": /* hex-encoded binary data */,
+        "application_nonce": /* hex-encoded binary data */,
+      },
+      ...
+    ],
+    ...
   ]
 }
 ```
 
 Verification:
 
-For all `N` entries in the `leaves`
-* `leaves[N].handshake[j].key = handshake_ratchet_key_[2*N]_[leaves[N].generation]`
-* `leaves[N].handshake[j].nonce = handshake_ratchet_nonce_[2*N]_[leaves[N].generation]`
-* `leaves[N].handshake[j].plaintext` represents a PublicMessage containing a
-  handshake message (Proposal or Commit) from leaf `N`
-* `leaves[N].handshake[j].ciphertext` represents a PrivateMessage object that
-  successfully decrypts to an MLSPlaintext equivalent to
-  `leaves[N].handshake[j].plaintext` using the keys for leaf `N` and generation
-  `leaves[N].generation`.
-* `leaves[N].application[j].key = application_ratchet_key_[2*N]_[leaves[N].generation]`
-* `leaves[N].application[j].nonce = application_ratchet_nonce_[2*N]_[leaves[N].generation]`
-* `leaves[N].application[j].plaintext` represents a MLSPlaintext containing
-  application data from leaf `N`
-* `leaves[N].application[j].ciphertext` represents a PrivateMessage object that
-  successfully decrypts to an MLSPlaintext equivalent to
-  `leaves[N].handshake[j].plaintext` using the keys for leaf `N` and generation
-  `leaves[N].generation`.
-* `sender_data_info.secret.key = sender_data_key(sender_data_secret, sender_data_info.ciphertext)`
-* `sender_data_info.secret.nonce = sender_data_nonce(sender_data_secret, sender_data_info.ciphertext)`
-
-The extra factor of 2 in `2*N` ensures that only chains rooted at leaf nodes are
-tested.  The definitions of `ratchet_key` and `ratchet_nonce` are in the
-[Encryption
-Keys](https://github.com/mlswg/mls-protocol/blob/master/draft-ietf-mls-protocol.md#encryption-keys)
-section of the specification.
+* `sender_data`:
+  * `key == sender_data_key(sender_data_secret, ciphertext)`
+  * `nonce == sender_data_nonce(sender_data_secret, ciphertext)`
+* Initialize a secret tree with a number of leaves equal to the number of
+  entries in the `leaves` array, with `encryption_secret` as the root secret
+* For each entry in `leaves`:
+  * For each entry in the array `leaves[i]`, verify that:
+    * `handshake_key = handshake_ratchet_key_[i]_[generation]`
+    * `handshake_nonce = handshake_ratchet_nonce_[i]_[generation]`
+    * `application_key = application_ratchet_key_[i]_[generation]`
+    * `application_nonce = application_ratchet_nonce_[i]_[generation]`
+
+The index `i` into the hash ratchets represents the leaf node with leaf index `i`. 
+
+## Message Protection
+
+Parameters:
+* Ciphersuite
+
+Format:
+
+``` text
+{
+  "cipher_suite": /* uint16 */,
```

Comment:
Is the sender index missing from this struct?

Reply:
My reason for including `sender` was much more benign -- it's simply useful to know the sender in advance s.t. we can set "myself" to not be the sender (our decryption function throws an error in such case).

Re. known-answer tests, let's not do it here, the test case is long enough as it is. FWIW, the way we do it is to have a "set_reuse_guard" function which is only compiled in tests (else it's internally chosen at random). Seems safe enough

Reply:
Yeah, I had some similar thoughts while implementing yesterday.  We could do something simple: Just have `n_leaves` always be 2, and always send from sender 1 (say).  That way, we can eliminate `n_leaves` from the test vector and avoid the issue you note.

Reply:
Sounds good!